streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -270,8 +270,11 @@ public void handleAssignment(final Map<TaskId, Set<TopicPartition>> activeTasks,
                 if (oldTask.isActive()) {
                     final Set<TopicPartition> partitions = standbyTasksToCreate.remove(oldTask.id());
                     newTask = standbyTaskCreator.createStandbyTaskFromActive((StreamTask) oldTask, partitions);
+                    cleanUpTaskProducer(oldTask, taskCloseExceptions);
                 } else {
                     oldTask.suspend(); // Only need to suspend transitioning standbys, actives should be suspended already
```

Comment:
This comment makes me a bit twitchy. Can we assert it instead? I mention it because I assume that active tasks also don't need to be committed because it should have happened already. Can we assert that as well?

Reply:
Just to be clear, we do assert/enforce this, but in the StreamTask and not in the TaskManager. At this point the TaskManager is actually completely agnostic to task state* and all assertions and branching based on state is internal to the Task implementation.

*except for in `handleRevocation`, where I just noticed we still filter the commit based on state, which is now unnecessary

Reply:
Your words sound good to me, but they are directly in contradiction to the code here, which skips suspending and committing active tasks because it assumes they have already happened. In other words, there is an assumption here that the active tasks are in some kind of "committed" state, while the standbys are in either "created" or "running".

If we're going to have a branch that explicitly assumes the task is already committed, then I'd like to verify it, otherwise experience says it will become false after refactoring, and we'd wind up trying to track down an IllegalStateException later on.

On the other hand, if these transitions are idempotent, we can just include them in both branches (or rather move them outside the conditional block).

Reply:
While `suspend` is technically idempotent, the safer thing is actually to _not_ also call it for active tasks here. But let me back up a bit and try to further clarify my earlier response. I realize I made this claim without any hint as to what specifically I was referring to:

By `we do assert/enforce this, but in the StreamTask` I meant that if we try to close a task without first suspending it, we will get an IllegalStateException in the `Task#close` method. The TaskManager is responsible for "managing" the operations on the task, but the task is ultimately responsible for verifying that its lifecycle makes sense and its state/transitions are valid. If we just blindly suspend everything here, it seems worse than getting an IllegalStateException because we lose the check that the task was prepared for close (ie suspended + committed) earlier during `handleRevocation`.

> I'd like to verify it, otherwise experience says it will become false after refactoring, and we'd wind up trying to track down an IllegalStateException later on

On that note, I'm slightly confused by your proposal:  what do you mean by "verify" if not to throw an IllegalStateException if it's not what's expected? Or do you just mean we'd have to track down an IllegalStateException slightly farther from the root cause

Reply:
Also, the `commit` case is trickier than the `suspend` as it's possible we actually _don't_ commit a task before closing it, if the commit or any commit preparation fails. So we can't just assert that the task is committed.

We can assert that we attempted to commit it by keeping track of tasks we tried to commit/revoke between `handleRevocation` and `handleAssignment`, but I'm quite confident that would quickly get out of control.
We could introduce a `#commitAttempted` method on the Task but that also seems to invite bugs
We could leave it up to the Task to make sure everything is done safely during the close procedure. What the task currently does is verify that no commit is needed if a clean close is attempted -- if we try to close clean but a commit _is_ still needed, it means the commit failed, and we can throw an exception to force the TM to closeDirty

Reply:
I think @vvcephei 's confusion comes from a change that we now require all tasks to be transited to `suspended` before transiting to `close`: previously, we allow e.g. a running task to be closed immediately and inside the task itself the logic actually did the "running -> suspend -> close" logic, i.e. it is totally agnostic to the TM. In a refactoring with eos-beta we changed it. So now the responsibility is kinda split between the two: TM needs to make sure the transition is valid and the task verifies it. By doing this we avoided the "pre-/post-" functions.

Reply:
Hi all, thanks for the discussion.

To start at the beginning, yes, I was advocating for throwing an IllegalStateException at the earliest possible moment when we can detect the illegal state. Right here in the code, we are making an invalid assumption. Namely, that if a task to be recycled is active, then it has already been suspended and committed, and if it is a standby, then it still needs to be suspended and committed. Why should this be true? Because some other code deep inside another set of nested conditionals thirty lines above looks like it does that right now? Experience says that this situation will not survive refactoring. We cannot test every branch, so we need to make assertions about the state being valid when it's in doubt.

We could make an argument that if this assumption becomes incorrect, than we'll throw an exception later on before any state becomes corrupted, which would be good. We could make a stronger argument that the exception we throw later on will be perfectly crystal clear about the cause and therefore we won't spend weeks poking around a flaky test or a user bug report trying to figure out what happened. But both of those arguments would depend on even further assumptions about stuff that may or may not happen elsewhere in the code base. The best thing to do at all times is validate potentially dangerous assumptions. This looks very much like a potentially dangerous assumption. I'm asking that we validate it.

Reply:
I won't hold up this PR on the point, but I just don't see how checking the state of the task here and throwing an exception if it's not in suspended at this random point in the TM code is any more clear than checking the state in the task and throwing an exception if it's not in suspended in `Task#close`

The latter makes it very clear what the problem is -- tasks need to be suspended before they are closed, end of story. The former seems to enforce suspension at an arbitrary point in the TM.

>Why should this be true? Because some other code deep inside another set of nested conditionals thirty lines above looks like it does that right now

This makes me wonder if I'm misinterpreting you, though, at least in part. Are you saying that it's fine to make this assumption thirty lines up when dealing with tasks that are actually being closed, but it's not ok to make this assumption here specifically when we are recycling tasks?

Reply:
Hey folks, taking a step back here regarding this, here are a few different principles we can exercise regarding the class hierarchy:

1) TM to be totally agnostic to the task's state, and the task's method would transit multiple state if necessary: e.g. if TM calls `close` on a running task, then task would first transit to suspended, and then to closed; if TM calls `suspend` on a suspended task, then it would be a no-op. In this case, then here TM could just blindly call `close` or `suspend` for closing / recycling tasks, and the Task's methods would handle state transition.

This would be ideal, but the current issue is that we need to commit post suspending a task today (i.e. handleRevocation would commit those suspended active tasks, while handleAssignment would commit those suspended standby tasks), and hence if we blindly suspend and commit a task, we may unnecessarily double committing a suspended task.

Right now to work around this issue, we choose to:

2) Letting TM to be aware of the task's state transition rules and tries to obey it. Similarly, we can also let TM to check the task's state, and then call suspend and commit conditionally -- at the moment, only standby tasks would not be suspended, so we would effectively end up the same logic as to only suspend standby tasks. So either checking `task.isActive` or `task.isSuspended` would work the same.

Personally I think a long term improvement could be, that we make preCommit / postCommit to be "idempotent" as well, e.g. if the task remembers since last `preCommit` we do not have processed any records, then a second call to `preCommit` would be a no-op and returns empty map, and similarly `postCommit` can also be a no-op if there's no processed records since its last checkpointing. And then in TM, we can blindly call:

```
task.suspend();
aggregatOffsets(task.prepreCommit());
if (!offsetsMap.isEmpty()) call sendOffsets();
postOffsets(task.postCommit());

if (isActive()) {
   recycle as standby;
} else {
   recycle as active;
}
```

And similarly for tasksToClose we can blindly call `suspend / preCommit / postCommit` first. WDYT?

Anyways, I think this discussion can continue in a follow-up PR (maybe I can incorporate whatever we've agreed on in the decoupling flushing/committing PR) and we can merge this one as-is.

Reply:
Thanks, all!

Thanks for adding the check, @ableegoldman . To answer your question, I do think there may be some kind of misunderstanding. Almost certainly, I failed to make myself clear.

I agree that the reason we enforce the valid transitions inside the task is so that the TM doesn't have to check every little thing. For example, that we call suspend in one place and close in another is not a big deal. As you pointed out, if we failed to call suspend, then the close call would let us know.

However, in this case, we are checking something. But what we're checking is only vaguely related to what we do next. In other words, `if (! isSuspended() ) suspend()` makes way more sense than `if ( isStandby() ) suspend()`. How are we supposed to know that `standby implies not suspended` at line 275 of a 700-line class? At least, with the log you added, if any quadrant of the implication is false, we'll have an error log telling us where it went wrong.

Note, I think I was especially uneasy because we're now also committing the task in this block, and "committed" or "not committed" isn't a checked state, which is how we wound up with the subtle bug that you're fixing here. I think your fix is fine; I just felt the pre-existing structure of the code needed improvement.

And thanks for the alternative proposals, @guozhangwang . I agree that either one of them would resolve my concern, and that we can take care of it in a follow-on PR.